### PR TITLE
Cipher Object in transit 

### PR DESCRIPTION
**Pull request by srinibas-zededa**
Pillar changes for handling object in transit cipher.
  -  review comments (#564)
  - code to handle for new proto changes, for certs API.

### CODE REVIEW

File: pkg/pillar/cmd/downloader/syncop.go [hunk 1]
```diff
@@ -292,18 +299,48 @@ func constructDatastoreContext(config types.DownloaderConfig, status *types.Down
...
+		cred.Password = dst.Password
+		return cred, nil
+	}
+	if !dst.IsValidCipher {
+		errStr := fmt.Sprintf("%s, Cipher Block is not ready", dst.Key())
+		return cred, errors.New(errStr)
+	}
+	clearData, err := tpmmgr.DecryptCipherBlock(dst.CipherBlock)
+	if err != nil {
+		log.Errorf("%s, dataStore CipherData decryption failed\n", dst.Key())
```
Comment (eriknordmark) [inline pkg/pillar/cmd/downloader/syncop.go]: please log %s err as well.

File: pkg/pillar/cmd/downloader/syncop.go [hunk 2]
```diff
@@ -292,18 +299,48 @@ func constructDatastoreContext(config types.DownloaderConfig, status *types.Down
...
+		return cred, err
+	}
+	if err := proto.Unmarshal(clearData, &cred); err != nil {
+		log.Errorf("%s, dataStore Credential unmarshall failed\n", dst.Key())
```
Comment (eriknordmark) [inline pkg/pillar/cmd/downloader/syncop.go]: Ditto

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 3]
```diff
@@ -870,6 +873,100 @@ func testEcdhAES() error {
...
+			if e != nil {
+				continue
+			}
+			certs = append(certs, c...)
+		}
+	}
+	if len(certs) == 0 {
+		return nil, errors.New("No X509 Certificate")
+	}
+	// use the first valid certificate in the chain
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go]: Only one cert makes sense, and you need to check that the ID aka truncated hash matches.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 4]
```diff
@@ -870,6 +873,100 @@ func testEcdhAES() error {
...
+			return []byte{}, err
+		}
+		return clearData, nil
+	}
+	return []byte{}, errors.New("Unsupported Encryption protocol")
+}
+
+func getControllerCertEcdhKey(cipherBlock types.CipherBlock) (*ecdsa.PublicKey, error) {
+	var ecdhPubKey *ecdsa.PublicKey
+	block := cipherBlock.ControllerCert
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go]: The certs will come from the cert device API since they don't change much and are very large.

Question is how the flow changes as you only get an ID aka truncated hash of the cert in the cipherBlock.
Are you planning to return an error if you don't have that ID and retry later after a GET of /certs to get the latest?

We need to work out that flow and structure now.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go (hunk 3 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go]: Sure. Will do.

File: pkg/pillar/cmd/downloader/syncop.go (hunk 2 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/downloader/syncop.go]: will do.

File: pkg/pillar/cmd/downloader/syncop.go (hunk 1 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/downloader/syncop.go]: Will do.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go (hunk 4 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go]: The cert API changes should trigger, reprocessing of existing cipherContexts/cipherBlocks.

Comment (eriknordmark) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go]: OK

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 5]
```diff
@@ -870,6 +873,100 @@ func testEcdhAES() error {
...
+	certs := []*x509.Certificate{}
+	for b, rest := pem.Decode(block); b != nil; b, rest = pem.Decode(rest) {
+		if b.Type == "CERTIFICATE" {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go]: How about an error if it is not a CERTIFICATE?

File: pkg/pillar/types/domainmgrtypes.go [hunk 6]
```diff
@@ -24,12 +24,16 @@ type DomainConfig struct {
 	Activate       bool   // Actually start the domU as opposed to prepare
 	AppNum         int    // From networking; makes the name unique
 	VmConfig
-	DiskConfigList    []DiskConfig
-	VifList           []VifInfo
-	IoAdapterList     []IoAdapter
+	DiskConfigList []DiskConfig
+	VifList        []VifInfo
+	IoAdapterList  []IoAdapter
+	// bo be deprecated, use CipherBlock instead
```
Comment (eriknordmark) [inline pkg/pillar/types/domainmgrtypes.go]: bo? XXX is useful to make it easier to find things to clean up.

File: pkg/pillar/types/zedroutertypes.go [hunk 7]
```diff
@@ -323,10 +323,15 @@ type CryptoBlock struct {
...
 	KeyScheme WifiKeySchemeType // such as WPA-PSK, WPA-EAP
-	Identity  string            // identity or username for WPA-EAP
-	Password  string            // string of pass phrase or password hash
-	Crypto    CryptoBlock       // encrypted block of items
-	Priority  int32
+	// to be deprecated, use CipherBlock instead
+	Identity string // identity or username for WPA-EAP
+	// to be deprecated, use CipherBlock instead
+	Password string      // string of pass phrase or password hash
+	Crypto   CryptoBlock // encrypted block of items
```
Comment (eriknordmark) [inline pkg/pillar/types/zedroutertypes.go]: CryptoBlock should presumably be expunged completely since you have defined CipherBlock for the same purpose

File: pkg/pillar/types/domainmgrtypes.go (hunk 6 above)
Comment (srinibas-zededa) [reply pkg/pillar/types/domainmgrtypes.go]: sure

File: pkg/pillar/types/zedroutertypes.go (hunk 7 above)
Comment (srinibas-zededa) [reply pkg/pillar/types/zedroutertypes.go]: will do.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go (hunk 5 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go]: it gets captured in the no certificate case, after the for loop.

File: pkg/pillar/types/cipherinfotypes.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+// Key :
+func (cipherBlock *CipherBlock) Key() string {
+	return cipherBlock.ID
+}
+
+// CredentialBlock : Credential Information
+type CredentialBlock struct {
+	Identity string
+	Password string
+}
```
Comment (eriknordmark) [inline pkg/pillar/types/cipherinfotypes.go]: This doesn't appear to be used.
You use zconfig.CredentialBlock but not this.

If it is intended to be used then it is OK. Otherwise do a separate PR to remove this.
